### PR TITLE
Add timeout and error messages

### PR DESCRIPTION
**Pull request by timgl**
## Changes

This adds a timeout message in case a query takes longer than 15 seconds. 

image

(To test this, run `bin/start-frontend` and `gunicorn posthog.wsgi -t 5 --workers=2`, then add `import time;time.sleep(7)` to insight.py def trends)

It also adds a nicer error message for queries
image

(Just add `assert 1 == 2` to an insight endpoint)

This also rejiggs quite a bit of the filters/results loading logic.

## Checklist

- [ ] All querysets/queries filter by Organization, by Team, and by User
- [ ] Django backend tests
- [ ] Jest frontend tests
- [ ] Cypress end-to-end tests

### CODE REVIEW

**Comment by timgl** [comment]
@paolodamico these are the error messages I mentioned to you, lmk if you want to change these up a bit

File: frontend/src/scenes/insights/EmptyStates.tsx
Comment (mariusandra) [inline frontend/src/scenes/insights/EmptyStates.tsx:44]: nice styles! :)

**Review by mariusandra — COMMENTED** [review]
I didn't run it, though it seems fine (still found a very useful nit though).
Could this also be extended to the funnels tab, but then with a much longer timeout, as they always load longer, at least in the postgres version.

File: frontend/src/scenes/insights/insightLogic.ts
Comment (EDsCODE) [inline frontend/src/scenes/insights/insightLogic.ts:92]: Hmm are we making a hard requirement now that all queries should run under 15 seconds? There are some queries that could probably cut closer to 20-30seconds

File: frontend/src/scenes/insights/trendsLogic.js
Comment (EDsCODE) [inline frontend/src/scenes/insights/trendsLogic.js:145]: Also on the Marius's point we could definitely use this message handling on all the insighs

**Review by EDsCODE — COMMENTED** [review]
Haven't QA'd yet but initial feedback

File: frontend/src/scenes/insights/insightLogic.ts
Comment (mariusandra) [reply frontend/src/scenes/insights/insightLogic.ts:92]: Two new points:
- Nit: This time should probably be a const somewhere
- The error message seems to indicate that we gave up, although if I understand correctly, the query is still running and might give results in some seconds anyway? If so, would be great to add a spinner somewhere to show it's still pending

Comment (timgl) [reply frontend/src/scenes/insights/insightLogic.ts:92]: The spinner will still be shown after 15 seconds so there is indication that it's still loading. I can reword the message if it's still going.

Agree re: const

**Review by paolodamico — CHANGES_REQUESTED** [review]
Didn't reviewed the code thoroughly as seems that @mariusandra has covered it. I did QA it. My comments:

- See below I caused the `insight` endpoint to fail, but the error message was not shown (see screenshot). Further it seemed that after this failed state, if I added/changed/removed filters, no further requests were tried (unsure if this was introduced in this PR). The "Add breakdown" & "Add filter" breakdown keep a loading spinner indeterminately.
    
- For recommendation 4 on error messages (CH migration) we show it even if the user has CH-enabled already. I know this is a very small subset of instances right now, but I think we'll forget to change this when we start to have a ton more adoption on CH.
- We should add some instrumentation, something like `insight timeout message shown` & `insight error message shown`.  And also `data-attr` and/or UTM tags to all the links.
- This definitely is an improvement towards a more robust product, and giving the sense of greater reliability. I'm torn if we should feature flag this. If we don't, we won't be able to measure impact on retention and other overall engagement metrics (even if experiments are a bit tainted from running too many things at once). If we do, we won't ship this significant improvement to all users ASAP.
- When showing the error screen, the default empty state of the graph is still kept.

- Agree with @EDsCODE that the timeout message looks like a terminal state too. Will take a pass at the UI tomorrow. Probably @lottiecoxon or @corywatilo could make a nicer illustration here to be more friendly, but I'll add something visual in the meantime to soften it a bit.

**Comment by timgl** [comment]
@paolodamico Thanks for QA'ing this well. I think I've fixed all the issues above (though do check, the logic is quite complicated).

Re: it looking like a terminal state, it's very easy now to pass through the loading state to the error message so we can make it look better.

**Comment by paolodamico** [comment]
Made some adjustments to the UI (see below). Probably @lottiecoxon could create nicer illustrations for this at some point in the future, but I think this is good for now.

**Error message**


**Timeout**

**Review by paolodamico — APPROVED** [review]
lgtm! all comments seem addressed, and after QA-ing, the messages seem to be shown as expected. Only thing I'm not clear on is that the timeout message is supposed to be shown if a query is taking a long time, **but is still in progress.** How does the message change after the full timeout of the request has elapsed and we know the query will not complete? 

I think we should maybe show a variation of the error message when the above happens. Holding off on the merge just to clarify that final point.

**Comment by timgl** [comment]
Awesome, thanks!

The timeout message will show after 15 seconds (or 30 seconds for funnels), or if we get an error message that isn't obviously a server error (there's no standard status code for timeouts), at which point the query is finished.

You can pass through `isLoading` to `<TimeOut />` to distinguish between those two states.

**Comment by mariusandra** [comment]
This is probably a bit too spinny:

**Comment by mariusandra** [comment]
Another point. When the backend errored (I just shut down the django server while waiting for a long query), the second spinner disappeared, but the top one still said that things are slow and "We are still working on it."

**Comment by mariusandra** [comment]
Last comment here. Long-running funnels timeout with no error. I just shut down celery and waited for the task to timeout. Once it did, the yellow "looks like things are a little slow" dialog was replaced with just nothing. It should probably show the error message instead.

Screenshot 2021-01-11 at 13 07 50

Somewhat related, the celery task to calculate the funnel will be "live" for [3 minutes](https://github.com/PostHog/posthog/blob/0b0e49c0c3bed34342351dedc0cfebb6dffb47a4/posthog/api/insight.py#L221), yet the frontend will keep polling for it for [2 minutes](https://github.com/PostHog/posthog/blob/5f07945cb5c887f8f58c56e7e0039fb340338c0e/frontend/src/scenes/funnels/funnelLogic.js#L12). The frontend code actually polls 120 times and waits a second plus however long it took to poll, so it's more like ~140sec. I'd replace this with a real timer that waits for 3 minutes and uses `window.performance.now()` to get an accurate timer (as `new Date().valueOf()` might change randomly if the user changes their clock).

I wrote `"live" for 3 minutes` in "quotes" because if I understand correctly, then if it takes 5 minutes to calculate the funnel, the celery task will run an extra two minutes and then discard the result. There's no actual timeout to cancel the expensive SQL query after 3 minutes. If someone requests a query that takes a few hours to complete and keeps refreshing the page often enough, they might just block all celery workers that way. 

Finally, a user in the users chat said that he'd be happy to wait for however long it takes to calculate the funnel. It would be cool to have a "keep waiting or cancel query" option, which could be used when you sometimes want to run the rare expensive query that takes 10min to complete. That's probably another task though.

**Comment by timgl** [comment]
@mariusandra fixed both those issues. I think it's quite hard to cancel sql queries mid-run. We can rely on statement_timeout set on the postgresql level (or on clickhouse). I'll leave the cancel queries for another PR.

**Comment by paolodamico** [comment]
Great stuff @mariusandra & @timgl! Just pushed one final adjustment to show a different icon after the query expires.

File: frontend/src/scenes/insights/insightLogic.ts
Comment (mariusandra) [inline frontend/src/scenes/insights/insightLogic.ts:19]: Was this accidental or on purpose? 

30sec is quite long though. AFAIK users start to get distracted and leave after ~10sec, so 15might still work, but anything more is too long. 3sec might be too short though.

**Comment by mariusandra** [comment]
- When the funnels load slow, the old funnel is still below with another spinner. Is this on purpose?
image

- When it times out, the view looks partial:
image

- The error "Your query took a long time to complete." is misleading. Reading it my thoughts go along these lines: "Took a long time to complete? So it completed? Show it to me!". This also is shown for network errors, which is even *wronger*.

**Comment by timgl** [comment]
Fixed all above. Happy to merge @mariusandra ?

**Review by mariusandra — CHANGES_REQUESTED** [review]
Nope, something's broken:

image

It keeps spinning endlessly. To replicate I just killed celery and then asked for a funnel.

**Comment by timgl** [comment]
ugh! should be good now

**Comment by mariusandra** [comment]
This is much better than what came before and seems to work, so I'll approve! However if I could nit again, then when the query just errors immediately, the text still says "took too long to complete". See screencast.

**Review by mariusandra — APPROVED** [review]

**Pull request merged**
